Sending error log (fixes #959)

krozycki opened the pull request:
Fixes #959 

Shows a modal dialog asking user whether he/she allows to send error logs.
Webgui to look over the collected logs: http://188.166.50.65:8081/db/logsdb/logs
After a fatal error, user is given a link to the webpage with the report that was sent.

:-)

AudriusButkevicius commented:
So I guess this wouldn't be very useful, because Fatal is usually something we might not be able to fix and we cannot continue without. For example, someone's config is unreadable, a port is already taken, etc.

I'd be more interested capturing panics, which can now probably done inside the monitor process, or by wrapping `syncthingMain()` with a deferred call to `recover()`

Also for centralized logging, I've used [Sentry](https://github.com/getsentry/sentry) before (which has a [Go client](http://godoc.org/github.com/getsentry/raven-go)) and it does a wonderful job, with logs per second, grouping similar messages and so on.

krozycki commented:
So how about just switching to using Sentry and capturing >= LevelWarn?

AudriusButkevicius commented:
Not sure, I'd like to hear @calmh's opinion.

calmh commented:
So, I'd say that when doing this, the *most* interesting thing would be panic logs. The ones that the monitor currently catches.

krozycki commented:
Great.
What about using Sentry vs using what is done now(mongodb + mongo-express)?

marclaporte commented on gui/index.html:
```diff
@@ -846,6 +853,25 @@ <h4 translate class="modal-title">Settings</h4>
+          <h4 translate class="modal-title">Allow Error Log Sending To Development Team?</h4>
+        </div>
+        <div class="modal-body">
+          <p translate>The anonymous error log will be send to development team everytime error occurs. It will be used to repair bugs and monitor problematic issues.</p>
```
send -> sent
everytime -> every time

Potential addition: "Syncthing is a community project and this is a good first step to get involved and help the project."

calmh commented:
Maybe? I have no personal experience of it, but I've heard good things.

krozycki commented:
I corrected spelling mistakes and rebased so that it can by merged again.

In my opinion we can try to pull the changes and see how it will act. We will know whether the collected logs will be useful and the whole idea worth taking care of. And if we think we'll lack some functionality it shouldn't be hard to switch to Sentry.

calmh commented:
From my point of view, this should be done from the monitor process that already monitors for panics. The panics are interesting - I'd love to get them all reported automatically. The fatal level log messages aren't really that exciting, most of them are just stuff the user should fix I think. Even if we do want to get them reported too (which I'm not *opposed to*), it should probably be done from the monitor process so that it has the context (i.e. preceding log lines) as they are not very informative by themselves...

AudriusButkevicius commented:
Tbh, I don't care about the fatal lines at all. These were added by us when we know there is nothing we can do and the user needs to get involved. Panics, yes, I am interested.

krozycki commented:
Aaaa, I confused fatal with panic somehow and thought it is ok. o.O

krozycki commented:
The monitor process is a separate process that just catches signals raised by syncthing process?

It would be nice to pass model and configuration options to monitor somehow, any ideas?

AudriusButkevicius commented:
So I think you can reuse the same functions which are already there.
I don't think you need the model, just the config?

calmh commented:
Config seems to be necessary to know if we should report errors, and if so where. Currently we load the config after separating from the monitor... Maybe that can be rearranged.

krozycki commented:
Model is needed in case we want to send the information that we send now with the usage reports. (platform, version, etc.)

AudriusButkevicius commented:
Platform, version etc are available as globals.

calmh commented:
Model is off limits; it's basically the whole of the non-monitor process. :)

krozycki commented:
I moved log sending from logger callback to monitor process where it watches for panics as you suggested.

I hope it is ok now.

Ps. Sorry for the time it took.

AudriusButkevicius commented:
So looks more or less good to me, a few things though:

1. I think we load the config as part of the monitor process, so we just need to pass it down to the monitor function which then can access the variable directly, and avoid this reading stdout stuff for clues.
2. Replace the hand-crafted json with a proper Go struct which we can serialize using the standard library.
Also, perhaps use separate fields for OS, ARCH just in case we filter stuff in the future.
3. ErrorLogSendingAccepted is quite cryptic, perhaps we should just call it PanicReports/PanicReporting/PanicReportingAccepted.
4. Would be nice to actually see how the output looks, because now it seems that it stores a document per line?
5. We potentially might need to annonimise stuff/use the same unique identifier which is not device ID as used in usage reporting to be able to distinguish between users.

krozycki commented:
Hi.

1) I think the config is being loaded later, in the main.go in the synthing process. That's also what calmh said a few messages above. ("Config seems to be necessary to know if we should report errors, and if so where. Currently we load the config after separating from the monitor... Maybe that can be rearranged.") So that's why I need to tell the monitor process somehow about the options.

2) Replaced, using json.Marshal right now. Also divided OS and ARCH.

3) Renamed.

4,5) I don't see what you mean to be honest. Why don't we just pull the changes and see how it behaves?

AudriusButkevicius commented:
1) Ok, I guess I would prefer to do it the right way (by loading the config in the monitor process), but if @calmh thinks it's fine, then it's fine with me too, as I am not too fused.
I feel that sooner or later we will need the config in the monitor, so it's just doing it sooner rather than later.

2,3) Looks good. I'll make a few more comments regarding capitalization and sentences on the actual lines. Sorry to be so picky.

4) I misunderstood what your code was doing, after checking your site I realized.
I suggest instead of sending `Error: line[:len(line)-1]` we should send `stdoutFirstLines + stdoutLastLines` which then has the relevant stack traces.

5) ... but as a side effect starts containing sensitive information (such as device ID's, IP's etc) which we should catch and replace using regex or some magical way. We should create another one of these https://github.com/syncthing/syncthing/blob/master/internal/config/config.go#L235 (or rename the existing one to UniqueObscureID or something) to be able to identify that it's the same person submitting the log, but at the same time maintain annonimity.

BONUS) We should also have a preview of what we are sending to the service same like we have for usage reporting just to be clear. This can be a static blob of text which we can hard code in the template, doesn't need to have any dynamic content.

AudriusButkevicius commented on cmd/syncthing/monitor.go:
```diff
@@ -181,6 +186,43 @@ func copyStderr(stderr io.ReadCloser, dst io.Writer) {
+						Os:          runtime.GOOS,
+						Arch:        runtime.GOARCH}
+					b, err := json.Marshal(report)
+					if err == nil {
```
You could invert the if and reduce the indentation
```
if err != nil {
  l.Warnln(err)
  continue
}

var jsonString...
```

krozycki replied:
I am afraid this will change the behaviour. There is an instruction after 'if' but in the 'for' that we do not want to ommit.

AudriusButkevicius replied:
Yes! Good point.

AudriusButkevicius commented on cmd/syncthing/monitor.go:
```diff
@@ -181,6 +186,43 @@ func copyStderr(stderr io.ReadCloser, dst io.Writer) {
 				}
 				stdoutMut.Unlock()
+
+				if optPanicReporting {
```
If we hit this branch, we should not print the the `Please create an issue at` line I guess.

krozycki replied:
I guess it doesn't make charm to print it either way so that we get pinged about a bug. After all if somebody creates an issue it means that it is a sever problem that affects his usage of Syncthing.

AudriusButkevicius commented on gui/index.html:
```diff
@@ -890,6 +890,13 @@ <h4 translate class="modal-title">Settings</h4>
+                <div class="form-group">
+                  <div class="checkbox">
+                    <label>
+                      <input id="PanicReporting" type="checkbox" ng-model="tmpOptions.panicReportingEnabled"> <span translate>Automatically Send Error Logs to Dev Team</span>
```
Automatically send anonymised fatal error logs to developers

AudriusButkevicius commented on gui/index.html:
```diff
@@ -911,6 +918,25 @@ <h4 translate class="modal-title">Settings</h4>
+    <div class="modal-dialog modal-lg">
+      <div class="modal-content">
+        <div class="modal-header alert alert-success">
+          <h4 translate class="modal-title">Allow Error Log Sending To Development Team?</h4>
```
Automatically send anonymised fatal error logs to developers?

AudriusButkevicius commented on gui/index.html:
```diff
@@ -911,6 +918,25 @@ <h4 translate class="modal-title">Settings</h4>
+          <h4 translate class="modal-title">Allow Error Log Sending To Development Team?</h4>
+        </div>
+        <div class="modal-body">
+          <p translate>The anonymous error log will be sent to development team every time error occurs. It will be used to repair bugs and monitor problematic issues. Syncthing is a community project and this is a good first step to get involved and help the project.</p>
```
repair -> fix

AudriusButkevicius commented on cmd/syncthing/main.go:
```diff
@@ -615,6 +615,11 @@ func syncthingMain() {
 		}
 	}
 
+	/* This is printed out to inform monitor process about options */
```
We try to use `//` for comments pretty much everywhere throughout the code.

krozycki commented:
Hmm, I don't know why it is failing in Jenkins now. :/

AudriusButkevicius commented:
Yeah jenkins is borked for me too.

krozycki commented:
And, regarding the things you mentioned. I guess we are blocked by not being able to filter the stdout attachted to the panic. We can not send it since it contains user data if we don't filter it and we also cannot show "Bonus" as we don't really know the final preview of the report.

calmh commented:
@st-jenkins retest this please

calmh commented on cmd/syncthing/monitor.go:
```diff
@@ -181,6 +186,43 @@ func copyStderr(stderr io.ReadCloser, dst io.Writer) {
+					if err == nil {
+						var jsonString = `docs=` + string(b)
+						// Send error log
+						resp, err := http.Post("http://188.166.50.65:27080/logsdb/logs/_insert", "application/x-www-form-urlencoded", strings.NewReader(jsonString))
```
So... What's the server side here? It seems to me that this should better be somewhat more generic, i.e. `http.Post("https://data.syncthing.net/panic")` and the receiving side can sort out what to do on the backend (i.e. push it into some database etc). Also, HTTPS and a DNS name. And hence push it just as a JSON blob of type `application/json` or something since all our other POST requests look like that?

krozycki replied:
I used sleepy.mongoose to get POST request and pass it to mongo. It needs either application/x-ww-form-urlencoded or multipart/* as the content type according to: http://stackoverflow.com/questions/17247347/why-do-post-file-uploads-in-gae-not-have-access-to-the-cgi-fieldstorage-methods I can switch this to application/json and a DNS name an leave you the server side if that's ok?

krozycki replied:
Server side I used: mongo + sleepy.mongoose mentioned above + mongo-express for viewing database. But I think it would be a good idea to set up Sentry under data.syncthing.net/panic and send this JSON blob of type application/json there.

calmh commented on gui/index.html:
```diff
@@ -911,6 +918,25 @@ <h4 translate class="modal-title">Settings</h4>
+    <div class="modal-dialog modal-lg">
+      <div class="modal-content">
+        <div class="modal-header alert alert-success">
+          <h4 translate class="modal-title">Automatically send anonymised fatal error logs to developers?</h4>
```
Title case, probably "Allow Anonymous Error Reporting?" to conform to the existing wording on usage reporting.

calmh commented on cmd/syncthing/monitor.go:
```diff
@@ -181,6 +186,43 @@ func copyStderr(stderr io.ReadCloser, dst io.Writer) {
+						Os          string
+						Arch        string
+					}
+					report := &Report{Error: line[:len(line)-1],
```
This won't actually help us very much, because this line will only contain `panic: runtime error: invalid memory address or nil pointer dereference` which tells us nothing about where it happened or why. We'll need both the preceding and following lines until the child syncthing process exits, i.e. like what gets printed to the panic log... Which makes it a little bit more tricky.

krozycki replied:
If only you could give me the regexp which will filter the stdout and make it anonymous (or any other solution), i'll change it right away. :-)

calmh replied:
Thing is, you want to capture every line from the one you have found, until syncthing exits, because that's the full backtrace of the crash. And then ideally the last few lines leading up to that, for context.

krozycki replied:
I see, but what if there are for example user's filenames or sth?

calmh replied:
Then that's probably part of the interesting stuff, as it leads up to why we crashed... But I see what you're getting at.

calmh commented:
Then that's probably part of the interesting stuff, as it leads up to why we crashed... But I see what you're getting at.

krozycki commented:
So what is the plan?

calmh commented:
Sorry, longest running pull request here. Can I suggest an alternate approach, if you're still interested in this? When starting up, if there are panic logs, ask the user to submit those (with preview), and then clear them out?
